File: core/src/com/mygdx/game/sprites/Tube.java
package com.mygdx.game.sprites;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector3;
import com.mygdx.game.states.GameStateManager;

import java.util.Random;

public class Tube {
    public Texture getTopTube() {
        return TopTube;
    }

    public Texture getBottomTube() {
        return BottomTube;
    }

    public Vector3 getPosTopTube() {
        return posTopTube;
    }

    public Vector3 getPosBottomTube() {
        return posBottomTube;
    }
    //width of the tube determined from measuring the png
    public static final float TUBE_WIDTH = 52;
    private static final float RANDRANGE = 150;
    private static final float GAP = 100;
    private static final float LOWESTOPENING = 120;
    private Texture TopTube, BottomTube;
    private Vector3 posTopTube, posBottomTube;
    private Random rand;
    // rectangles for collision detection
    private Rectangle boundsTop, boundsBot;


    public Tube(float x){
        //getting the image for the top and bottom tubes
        TopTube = new Texture("toptube.png");
        BottomTube = new Texture("bottomtube.png");
        // creating a random number for randomising positions of the tubes
        rand = new Random();

        // randomising the position of the top tube and bottom tubes
        posTopTube = new Vector3(x, rand.nextFloat()*RANDRANGE + GAP + LOWESTOPENING, 0);
        posBottomTube = new Vector3(x, posTopTube.y - GAP - BottomTube.getHeight(), 0);

        //creating the collision boundaries for the top and bottom tubes
        boundsTop = new Rectangle(posTopTube.x, posTopTube.y, TopTube.getWidth(), TopTube.getHeight());
        boundsBot = new Rectangle(posBottomTube.x, posBottomTube.y, BottomTube.getWidth(), BottomTube.getHeight());

    }

    public void reposition(float x){
        // randomly repositioning the top and bottom tubes
        posTopTube.set(x, rand.nextFloat()*RANDRANGE + GAP + LOWESTOPENING, 0);
        posBottomTube.set(x, posTopTube.y - GAP - BottomTube.getHeight(), 0);
        // setting rectangle boundaries to share the same positions as the tubes
        boundsTop.setPosition(posTopTube.x, posTopTube.y);
        boundsBot.setPosition(posBottomTube.x, posBottomTube.y);
    }

    public boolean collision(Rectangle user){
        return user.overlaps(boundsTop) || user.overlaps(boundsBot);
    }
}

File: core/src/com/mygdx/game/sprites/Tube.kt
package com.mygdx.game.sprites

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector3
import java.util.*

class Tube(x: Float) {

    val topTube: Texture
    val bottomTube: Texture
    val posTopTube: Vector3
    val posBottomTube: Vector3
    private val rand: Random

    // rectangles for collision detection
    private val boundsTop: Rectangle
    private val boundsBot: Rectangle
    fun reposition(x: Float) {
        // randomly repositioning the top and bottom tubes
        posTopTube[x, rand.nextFloat() * RANDRANGE + GAP + LOWESTOPENING] = 0f
        posBottomTube[x, posTopTube.y - GAP - bottomTube.height] = 0f
        // setting rectangle boundaries to share the same positions as the tubes
        boundsTop.setPosition(posTopTube.x, posTopTube.y)
        boundsBot.setPosition(posBottomTube.x, posBottomTube.y)
    }

    fun colides(user: Rectangle): Boolean {
        return user.overlaps(boundsTop) || user.overlaps(boundsBot)
    }

    fun dispose() {
        topTube.dispose()
        bottomTube.dispose()
    }

    companion object {
        //width of the tube determined from measuring the png
        const val TUBE_WIDTH = 52f
        private const val RANDRANGE = 150f
        private const val GAP = 100f
        private const val LOWESTOPENING = 120f
    }

    init {
        //getting the image for the top and bottom tubes
        topTube = Texture("toptube.png")
        bottomTube = Texture("bottomtube.png")
        // creating a random number for randomising positions of the tubes
        rand = Random()

        // randomising the position of the top tube and bottom tubes
        posTopTube = Vector3(x, rand.nextFloat() * RANDRANGE + GAP + LOWESTOPENING, 0f)
        posBottomTube = Vector3(x, posTopTube.y - GAP - bottomTube.height, 0f)

        //creating the collision boundaries for the top and bottom tubes
        boundsTop = Rectangle(posTopTube.x, posTopTube.y, topTube.width.toFloat(), topTube.height.toFloat())
        boundsBot = Rectangle(posBottomTube.x, posBottomTube.y, bottomTube.width.toFloat(), bottomTube.height.toFloat())
    }
}